runtime/lua/vim/treesitter/highlighter.lua

```diff
@@ -382,21 +380,35 @@ function TSHighlighter._on_spell_nav(_, _, buf, srow, _, erow, _)
 end
 
 ---@private
----@param _win integer
 ---@param buf integer
 ---@param topline integer
 ---@param botline integer
-function TSHighlighter._on_win(_, _win, buf, topline, botline)
+function TSHighlighter._on_win(_, _, buf, topline, botline)
   local self = TSHighlighter.active[buf]
-  if not self then
+  if not self or self.parsing then
+    return false
+  end
+  self.parsing = self.tree:parse({ topline, botline + 1 }, function(trees)
+    if trees then
+      self:_async_parse_callback(buf)
```

Comment:
If the buffer times out asynchronously, then the parsing flag is never reset and buffer is never highlighted. Even when parsed successfully (by someone other than the highlighter).
I think we should reset the `parsing` flag. But only if the changetick is different (so that the buffer isn't reparsed if it timed out & hasn't changed).

Reply:
Does it work if `prepare_callback_states()` checks if individual trees are valid instead of using the flag?

Also IIUC, one unparsed injection currently makes the highlighter skip the window.

Reply:
Maybe? Seems to work. I'm not familiar enough with treesitter to tell if this makes sense or would fail somewhere:
```diff
diff --git a/runtime/lua/vim/treesitter/highlighter.lua b/runtime/lua/vim/treesitter/highlighter.lua
index c8531a0285..3582dff6c0 100644
--- a/runtime/lua/vim/treesitter/highlighter.lua
+++ b/runtime/lua/vim/treesitter/highlighter.lua
@@ -69,7 +69,6 @@ end
 ---@field private _queries table<string,vim.treesitter.highlighter.Query>
 ---@field tree vim.treesitter.LanguageTree
 ---@field private redraw_count integer
----@field private parsing boolean true if we are parsing asyncronously
 local TSHighlighter = {
   active = {},
 }
@@ -169,9 +168,12 @@ end
 ---@private
 function TSHighlighter:prepare_highlight_states(srow, erow)
   self._highlight_states = {}
+  local has_valid = false
 
   self.tree:for_each_tree(function(tstree, tree)
-    if not tstree then
+    local is_valid = tstree and tree:is_valid(true)
+    has_valid = has_valid or is_valid
+    if not tstree or not is_valid then
       return
     end
 
@@ -199,6 +201,8 @@ function TSHighlighter:prepare_highlight_states(srow, erow)
       highlighter_query = highlighter_query,
     })
   end)
+
+  return has_valid
 end
 
 ---@param fn fun(state: vim.treesitter.highlighter.State)
@@ -388,25 +392,18 @@ function TSHighlighter._on_win(_, _, buf, topline, botline)
   if not self or self.parsing then
     return false
   end
-  self.parsing = self.tree:parse({ topline, botline + 1 }, function(trees)
+  self.tree:parse({ topline, botline + 1 }, function(trees)
     if trees then
       self:_async_parse_callback(buf)
     end
-  end) --[[@as boolean]]
-  if self.parsing then
-    return false
-  end
+  end)
   self.redraw_count = self.redraw_count + 1
-  self:prepare_highlight_states(topline, botline)
-  return true
+  return self:prepare_highlight_states(topline, botline)
 end
 
 ---@param buf integer
 function TSHighlighter:_async_parse_callback(buf)
-  if self.parsing then
-    self.parsing = false
-    api.nvim__redraw({ buf = buf, valid = false, flush = false })
-  end
+  api.nvim__redraw({ buf = buf, valid = false, flush = false })
 end
 
 api.nvim_set_decoration_provider(ns, {
```

Reply:
Does it cause double redraw for 2 windows?

I noticed it redraws twice when I was testing if the condition on `nvim__redraw()` is needed

Reply:
Oh yeah we still shouldn't call nvim__redraw in the synchronous case.

Reply:
During the same testing, I also noticed that in big files, it for some reason doesn't redraw all the lines when the tree is invalidated if `nvim__redraw()` isn't called in `on_win()`.

I.e. in this file:

```lua
vim.api.nvim_feedkeys('a', '', false)
local _ = "
vim.api.nvim_feedkeys('a', '', false) -- repeated many times
```

the redraw after deleting the `local _ = "` line still displays the lines below as string until the parse completes/cursor updates them.

Both with and without https://github.com/neovim/neovim/pull/31710

Reply:
For the first redraw that happens, I think that is expected. Deleting a line normally uses the "scroll redraw" path, where lines below the changes are scrolled up without actually calling `win_line()`. I don't think tree invalidation happens before the first redraw?

Reply:
Yeah I couldn't make sense of it either while first working with the tree sitter module working on the conceal_lines PR.

But if there is an issue there, it's separate from this PR right? Or are you seeing changed behavior?

Reply:
Thanks for the discussion, I will apply the above patch. Is there any other work to be done here before I start rebasing?

Reply:
I think other things are outside the scope

Reply:
Just noticed the original bug (no highlighting when timed out) is still present

Reply:
Isn't that intended? If timeout was reached how can highlighting be applied

It still applies later with subsequent parses or calls to treesitter.start()

Reply:
It does? When I was checking, it stopped highlighting amd never resumed, even when I removed most of the file. I assumed that the `parsing` flag is never reset if it times out.

Reply:
It should; a timeout still calls the `on_parse` callback, which resets `parsing`.

Reply:
If it times out, `trees` are `nil` and `_async_parse_callback()`, which resets the `parsing` is never called. `parsing` is only set in 2 places, and the second place only runs if `parsing` is false.

Reply:
If it times out, sync or async, `trees` are `nil` and `_async_parse_callback()`, which resets `parsing`, is never called. `parsing` is set only in 2 places (callback and `_on_win()`), and `_on_win()` returns early if `parsing` is `false`.

Reply:
Oh, I missed the `trees` guard in the highlighter's `on_parse` (do we need the separate  `_async_parse_callback()` method?). I guess that should be removed. Though that does beg to question what the point of the timeout is, because as you say the next redraw we simply try again.

Reply:
For some reason changing 
```lua
if not tstree then
```
to
```lua
if not tstree or not tree:is_valid(true) then
```
breaks injection highlighting, at least in `:h treesitter.txt`. Even on master, where this is the only delta (no async highlighting).

I don't think we need it though, especially after the patch to prevent the segfault when iterating captures after a parser has been reset

Reply:
It starts working after scrolling through the entire file. `LanguageTree` is not valid until all ranges are parsed. I also think it is not needed.